protocols/primary-backup/src/main/java/io/atomix/protocols/backup/roles/AsynchronousReplicator.java

```diff
@@ -110,8 +123,7 @@ private void backup() {
           context.currentIndex(),
           batch);
       log.trace("Sending {} to {}", request, nodeId);
-      context.protocol().backup(nodeId, request);
-      lastSent = System.currentTimeMillis();
+      return context.protocol().backup(nodeId, request).thenRun(() -> lastSent = System.currentTimeMillis());
```

Comment:
Thinking about this more I think it would be safer to drop the backupFuture field and set the timestamp before invoking backup.

Reply:
NettyMessagingService don‘t guarantee message to be sent on order,  if invoke `replicate` too quickly, backupRole may receive  two out-of-order `BackupRequest`, maybe I think need backupFuture to guarantee the order here

Reply:
It should keep the order now with our change that collects all the futures.

Reply:
sorry, I don't understand what you mean, 
future of `maybeBackup` is not collected.

Reply:
And if we set the timestamp before returning the actual cf it shouldn't be executed in parallel.

Reply:
but we cannot assume the message will be sent within 50ms, 
another case 
*  `maybeBackup` -> `backup` is invoked,  
* 100 `BackupOperation`s in a short time, invoke `backup` again

Reply:
@tanbangcheng how about something like this? https://gist.github.com/johnou/55bf838142b38695830f7941368ce9fb

Reply:
It makes sense to me to record the timestamp after the response. Correctness doesn't depend on storing it prior to the request.